### PR TITLE
8323183: ClassFile API performance improvements

### PR DESCRIPTION
**Pull request by asotona**
ClassFile API performance related improvements have been separated from #17121 into this PR.

These improvements are important to minimize performance regression of
8294961: Convert java.base/java.lang.reflect.ProxyGenerator to use the Classfile API to generate proxy classes #17121

Please review.

Thanks,
Adam

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323183](https://bugs.openjdk.org/browse/JDK-8323183): ClassFile API performance improvements (**Enhancement** - P4)

### Reviewers
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17306/head:pull/17306` \
`$ git checkout pull/17306`

Update a local copy of the PR: \
`$ git checkout pull/17306` \
`$ git pull https://git.openjdk.org/jdk.git pull/17306/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17306`

View PR using the GUI difftool: \
`$ git pr show -t 17306`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17306.diff">https://git.openjdk.org/jdk/pull/17306.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17306#issuecomment-1881034736)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java [hunk 1]
```diff
@@ -343,6 +335,51 @@ public StackCounter(LabelContext labelContext,
...
+                    inArray = false;
+                }
+                case '[' -> {
+                    inArray = true;
+                }
+                case ')' -> {
+                    if (cur < end) {
+                        if (subReturn) {
+                            return switch (descriptor.charAt(cur++)) {
+                                case 'Z', 'B', 'C', 'S', 'I', 'F', '[', 'L' -> count - 1;
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java:367]: What if the return type is an array?

Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java:367]: Ah nvm, I see you subtract -1 there

File: src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java [hunk 2]
```diff
@@ -343,6 +335,51 @@ public StackCounter(LabelContext labelContext,
         maxStack += rets * maxStack;
     }
 
+    private static int countMethodStack(Utf8Entry descriptor, boolean subReturn) {
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java:338]: Since we use this for both locals and stack, perhaps "countMethodSlots" would be better? And maybe the "subReturn" could be a flag to switch between locals/stack

File: src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java [hunk 3]
```diff
@@ -377,6 +414,6 @@ private void error(String msg) {
                 msg,
                 bcs.bci,
                 methodName,
-                methodDesc.displayDescriptor()));
+                MethodTypeDesc.ofDescriptor(methodDesc.stringValue()).displayDescriptor()));
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java:417]: This seems unrelated?

**Comment by liach** [comment]
You need to update the slot counting from `DirectCodeBuilder` and `StackMapDecoder` to fully avoid creating any MethodTypeDesc.

File: src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java (hunk 2 above)
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java]: I've changes the name to the suggested "countMethodSlots" and "subReturn" to "asStackDelta" , thanks.

**Comment by asotona** [comment]
> You need to update the slot counting from `DirectCodeBuilder` and `StackMapDecoder` to fully avoid creating any MethodTypeDesc.

It would be good to avoid all bottlenecks, however not all of them have equal effect.
This patch avoids MTD construction from for every invocation instruction. We may cache the symbol in the relevant Utf8Entry, however it still means to create each MTD at least once. "Hairy" methods doing a lot of various invocations are the most affected.

File: src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java [hunk 4]
```diff
@@ -343,6 +335,51 @@ public StackCounter(LabelContext labelContext,
...
+                case 'Z', 'B', 'C', 'S', 'I', 'F' -> {
+                    count++;
+                    inArray = false;
+                }
+                case 'J', 'D' -> {
+                    count += inArray ? 1 : 2;
+                    inArray = false;
+                }
+                case 'L' -> {
+                    while (cur < end && descriptor.charAt(cur++) != ';');
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java:356]:
```suggestion
                    cur = descriptor.indexOf(';', cur) + 1;
                    if (cur == 0)
                        throw new IllegalArgumentException("Bad method descriptor: " + descriptor);
```

**Comment by asotona** [comment]
StackMapGenerator benchmark is heavily benefiting from symbols caching for repeated execution.
I've created new CodeAttributeTools::benchmarkStackMapsGenerator and CodeAttributeTools::benchmarkStackCounter with focus on single use of each parsed model to eliminate background caching effect.
Then I've compared version with custom method parameter slots counting and version with symbols cached in NaT entries.
There is insignificant difference between these two approaches.
Explanation may by that repeated uses of the same NaT entry (of the called method) compensate the initial parsing cost. 

Based on these results I've decided to revert custom method parameter slots counting method and apply the previously proposed solution.

File: src/java.base/share/classes/jdk/internal/classfile/impl/StackMapDecoder.java [hunk 5]
```diff
@@ -80,7 +80,8 @@ public static List<VerificationTypeInfo> initFrameLocals(ClassEntry thisClass, S
         } else {
             vtis = new VerificationTypeInfo[methodType.parameterCount()];
         }
-        for(var arg : methodType.parameterList()) {
+        for(int pi = 0; pi < methodType.parameterCount(); pi++) {
```
Comment (turbanoff) [inline src/java.base/share/classes/jdk/internal/classfile/impl/StackMapDecoder.java:83]:
```suggestion
        for (int pi = 0; pi < methodType.parameterCount(); pi++) {
```

**Comment by asotona** [comment]
Please review.

Thank you!

**Review by cl4es — CHANGES_REQUESTED** [review]
Improvement looks good, but the microbenchmark needs to sink new instances into a blackhole to make sure we measure the right thing.

File: src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java [hunk 6]
```diff
@@ -25,26 +25,26 @@
...
 import java.util.BitSet;
-import java.util.LinkedHashMap;
 import java.util.List;
-import java.util.stream.Collectors;
 import java.lang.classfile.TypeKind;
 import java.lang.classfile.constantpool.ConstantDynamicEntry;
 import java.lang.classfile.constantpool.DynamicConstantPoolEntry;
 import java.lang.classfile.constantpool.MemberRefEntry;
 import static java.lang.classfile.ClassFile.*;
+import java.lang.constant.MethodTypeDesc;
```
[the remark this answers is not in the document]
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/impl/StackCounter.java:32]: Fixed, thanks.

File: test/micro/org/openjdk/bench/jdk/classfile/CodeAttributeTools.java [hunk 7]
```diff
@@ -105,19 +104,28 @@ public void setup() throws IOException {
     }
 
     @Benchmark
-    public void benchmark() {
-        if (it == null || !it.hasNext())
-            it = data.iterator();
-        var d = it.next();
-        new StackMapGenerator(
+    public void benchmarkStackMapsGenerator() {
+        for (var d : data) new StackMapGenerator(
```
[the remark this answers is not in the document]
Comment (asotona) [reply test/micro/org/openjdk/bench/jdk/classfile/CodeAttributeTools.java]: Fixed, thanks.
